### PR TITLE
 [CGP] Drop poison-generating flags after hoisting

### LINKED ISSUE #90380: [RISCV] Miscompile with -O3
Reduced test case:
```
#include "csmith.h"
uint8_t a = 55;
int32_t b;
int32_t c[4];
uint32_t e[4];
void f() {
  int32_t *o = &c[1];
  *o = 223;
  if (safe_mod_func_uint8_t_u_u(safe_sub_func_uint8_t_u_u(b, 1), a |= *o)) {
    e[0] = ++a;
  }
}
int main() {
  b = 0;
  f();
  printf("%d\n", a);
  return 0;
}
```
```
> bin/clang -O0 --target=riscv64-linux-gnu test.c -I/usr/include/csmith -w
> qemu-riscv64 -L /usr/riscv64-linux-gnu/ a.out
255
> bin/clang -O3 --target=riscv64-linux-gnu test.c -I/usr/include/csmith -w
> qemu-riscv64 -L /usr/riscv64-linux-gnu/ a.out
224
```
llvm version: d6c4ebbf78c67cd298f6ba2da6c06e2362f4d0f4

### PR DESCRIPTION
**Pull request by dtcxzyw**
See the following case:
```
define i8 @src1(i8 %x) {
entry:
  %cmp = icmp eq i8 %x, -1
  br i1 %cmp, label %exit, label %if.then

if.then:
  %inc = add nuw nsw i8 %x, 1
  br label %exit

exit:
  %retval = phi i8 [ %inc, %if.then ], [ -1, %entry ]
  ret i8 %retval
}

define i8 @tgt1(i8 %x) {
entry:
  %inc = add nuw nsw i8 %x, 1
  %0 = icmp eq i8 %inc, 0
  br i1 %0, label %exit, label %if.then

if.then:                                          ; preds = %entry
  br label %exit

exit:                                             ; preds = %if.then, %entry
  %retval = phi i8 [ %inc, %if.then ], [ -1, %entry ]
  ret i8 %retval
}
```
`optimizeBranch` converts `icmp eq X, -1` into cmp to zero on RISC-V and hoists the add into the entry block. Poison-generating flags should be dropped as they don't still hold.

Proof: https://alive2.llvm.org/ce/z/sP7mvK
Fixes https://github.com/llvm/llvm-project/issues/90380

### CODE REVIEW

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 1]
```diff
@@ -8268,8 +8268,10 @@ static bool optimizeBranch(BranchInst *Branch, const TargetLowering &TLI,
     if (CmpC.isPowerOf2() && Cmp->getPredicate() == ICmpInst::ICMP_ULT &&
         match(UI, m_Shr(m_Specific(X), m_SpecificInt(CmpC.logBase2())))) {
       IRBuilder<> Builder(Branch);
-      if (UI->getParent() != Branch->getParent())
+      if (UI->getParent() != Branch->getParent()) {
```
Comment (nikic) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:8271]: Shouldn't the poison flags always be dropped? I believe the relevant change here is that we introduce a new use of the instruction, not that we move it.

Comment (dtcxzyw) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp]: Done.

**Comment by antoniofrighetto** [comment]
I just wonder if there are possibly other places in CGP where we should drop the flags, but I assume that, unless we have a test case showing that explicitly, it's fine as it is.

**Review by nikic — APPROVED** [review]
LGTM, but is it possible to add these tests in llvm/test/Transforms/CodeGenPrepare/ARM/branch-on-zero.ll, which has the other tests for this transform?

**Pull request merged**